### PR TITLE
Allow override of CA certs for cert from client based on SNI server name sent by client.

### PR DESCRIPTION
**Pull request by ywkaras**
Requires OpenSSL interface 1.0.2.

### CODE REVIEW

**Comment by ywkaras** [comment]
Doc changes and Au testing pending.

**Comment by ywkaras** [comment]
[approve ci docs]

**Comment by shinrich** [comment]
Doesn't this require openssl 1.1.1 or later for the feature to be present?

File: doc/admin-guide/files/sni.yaml.en.rst [hunk 1]
```diff
@@ -69,11 +69,22 @@ verify_server_properties  One of the values :code:`NONE`, :code:`SIGNATURE`, :co
...
+verify_client_ca_certs    Specifies an alternate set of certificate authority certs to use to verify the
+                          client cert.  The value must be either a file path, or a nested set of key /
+                          value pairs.  If the value is a file path, it must specify a file containing the
+                          CA certs.  Otherwise, there should be up to two nested pairs.  The possible keys
+                          are ``file`` and ``dir``.  The value for ``file`` must be a file path for a file
+                          containing CA certs.  The value for ``dir`` must be a file path for an OpenSSL
+                          X509 hashed directory containing CA certs.  If a given file path does not being
+                          with ``/`` , it must be relative to the |TS| configuration directory.
+                          ``verify_client_ca_certs`` can only be used with capbilities provided by
+                          OpenSSL 1.0.2 or later.
```
Comment (shinrich) [inline doc/admin-guide/files/sni.yaml.en.rst:86]: Doesn't this feature require openssl 1.1.1?

File: doc/admin-guide/files/sni.yaml.en.rst [hunk 2]
```diff
@@ -69,11 +69,22 @@ verify_server_properties  One of the values :code:`NONE`, :code:`SIGNATURE`, :co
 verify_client             One of the values :code:`NONE`, :code:`MODERATE`, or :code:`STRICT`.
                           If ``NONE`` is specified, |TS| requests no certificate.  If ``MODERATE`` is specified
                           |TS| will verify a certificate that is presented by the client, but it will not
-                          fail the TLS handshake if new certificate is presented.  If ``STRICT`` is specified
+                          fail the TLS handshake if no certificate is presented.  If ``STRICT`` is specified
                           the client must resent a certificate during the TLS handshake.
 
                           By default this is :ts:cv:`proxy.config.ssl.client.certification_level`.
 
+verify_client_ca_certs    Specifies an alternate set of certificate authority certs to use to verify the
```
Comment (shinrich) [inline doc/admin-guide/files/sni.yaml.en.rst:77]: I assume this argument is optional.

File: doc/admin-guide/files/sni.yaml.en.rst (hunk 1 above)
Comment (shinrich) [reply doc/admin-guide/files/sni.yaml.en.rst:86]: Hmm.  Just checked the openssl docs.  Guess I misremembered.  SSL_set1_verify_cert_store appears to come in 1.0.2 or better.

File: doc/admin-guide/files/sni.yaml.en.rst (hunk 2 above)
Comment (ywkaras) [reply doc/admin-guide/files/sni.yaml.en.rst:77]: Yes.  Should it explicitly say that?

File: tests/gold_tests/tls/ssl/openssl.cnf [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+[ req ]
+distinguished_name = req_distinguished_name
+
+[ req_distinguished_name ]
+countryName                     = Country Name (2 letter code)
+countryName_default             = US
+countryName_min                 = 2
+countryName_max                 = 2
+stateOrProvinceName             = Michigan
+localityName                    = Hell
```
Comment (randall) [inline tests/gold_tests/tls/ssl/openssl.cnf:32]: nice!

Comment (ywkaras) [reply tests/gold_tests/tls/ssl/openssl.cnf:32]: https://www.google.com/maps/place/Hell,+MI+48169/@42.434735,-83.9937548,15z/data=!3m1!4b1!4m5!3m4!1s0x882334578757ad71:0xd45079770956618a!8m2!3d42.43472!4d-83.985

**Comment by ywkaras** [comment]
I don't think this is urgent enough to backport to 9.0 but it should go into 9.1.

**Comment by zwoop** [comment]
Cherry-picked to v9.0.x branch.

**Pull request merged**